R/ranges_helpers.R

```diff
@@ -143,35 +143,103 @@ asTX <- function(grl, reference) {
   return(pmapToTranscripts(grl, reference))
 }
 
-#' Faster more secure version of mapFromTranscripts
-#'
-#' Fixes a bug in function, and should have 10x speedup
-#' Also removes hit column for you
+#' Faster pmapFromTranscript
 #'
-#' @param ranges IRanges of ranges within grl
-#' @param grl the "transcripts" that contain ranges, GRangesList
-#' @param indices integer vector of which index of grl ranges are from:
-#' (c(1,1,2)) means first two ranges are from grl[1], third from grl[2])
-#' @return A GrangesList of ranges mapped from transcripts, names are
-#'  re-used from the grl (the transcripts).
-#' @family ExtendGenomicRanges
+#' This version tries to fix the shortcommings of GenomicFeature's version.
+#' Much faster and uses less memory.
+#' Implemented as dynamic program optimized c++ code.
+#' @param x IRangesList/IRanges to map to genomic coordinates
+#' @param transcripts a GRangesList to map agains
+#' @param removeEmpty a logical, remove non hit exons, else they are set
+#'  to 0.
+#' @return a GRangesList of mapped reads
+#' @export
+#' @examples
+#' ranges <- IRanges(start = c( 5, 6), end = c(10, 10))
+#' seqnames = rep("chr1", 2)
+#' strands = rep("-", 2)
+#' grl <- split(GRanges(seqnames, IRanges(c(85, 70), c(89, 82)), strands),
+#'              c(1,1))
+#' ranges <- split(ranges, c(1,1)) # both should be mapped to transcript 1
+#' x <- ranges
+#' transcripts <- grl
+#' pmapFromTranscriptF(ranges, grl, TRUE)
 #'
-pmapFromTranscriptF <- function(ranges, grl, indices) {
-  if (length(ranges) != length(indices)) stop("length of ranges != indices")
-  names <- names(grl)
-  names(grl) <- NULL
-
-  genomicCoordinates <- pmapFromTranscripts(x = ranges,
-                                            transcripts = grl[indices])
-  names(genomicCoordinates) <- names[indices]
-
-  genomicCoordinates <- genomicCoordinates[width(genomicCoordinates) > 0L]
-  a <- unlistGrl(genomicCoordinates)
-  a$hit <- NULL
-  return(relist(a, genomicCoordinates))
+pmapFromTranscriptF <- function(x, transcripts, removeEmpty = FALSE) {
+  if(is(x, "IRangesList")) {
+    # Create Ranges object from orf scanner result
+    x = unlist(x, use.names = TRUE)
+    indices <- strtoi(names(x))
+    names(x) <- NULL
+  } else if(is(x, "IRanges")) {
+    if(is.null(names(x))) {
+      indices <- seq.int(1,length(x))
+    } else {
+      indices <- strtoi(names(x))
+    }
+    names(x) <- NULL
+  } else stop("x must either be IRanges or IRangesList")
+  if(!is.logical(removeEmpty)) stop("removeEmpty must be logical")
+  removeEmpty <- as.logical(removeEmpty)
+  if(max(indices) > length(transcripts)) stop("invalid names of IRanges")
+  if (length(x) != length(indices)) stop("length of ranges != indices")
+
+  tx <- ranges(transcripts)
+  names <- names(tx)
+  names(tx) <- NULL
+  tx <- tx[indices]
+  if (!all(width(x) <=  as.integer(sum(width(tx))))) {
+    stop("Invalid ranges to map, check them. One is bigger than its reference")
```

Comment:
Invalid ranges to map, at least one of your ranges is bigger than its reference.